### PR TITLE
Re-introduce the scheduler performance improvements

### PR DESCRIPTION
**Pull request by wolfs**
Re-introduce the scheduler performance improvements, now
that https://github.com/gradle/gradle/pull/10050 has
been fixed.

This reverts commit 18b6c927b38e48eed57a5db28f9939ceb84de3d4.

Fixes https://github.com/gradle/gradle/issues/10094.

### CODE REVIEW

**Comment by wolfs** [comment]
The changes show a 3% improvement in [assembleDebug on santaTrackerAndroidBuild](https://builds.gradle.org/repository/download/Gradle_Check_PerformanceTestCoordinator/25535544:id/report-performance-performance-tests.zip%21/report/tests/assembleDebug-on-santaTrackerAndroidBuild.html?redirectSupported=false).

File: subprojects/core/src/main/java/org/gradle/execution/plan/DefaultExecutionPlan.java [hunk 1]
```diff
@@ -625,12 +630,22 @@ public Node selectNext(WorkerLeaseRegistry.WorkerLease workerLease, ResourceLock
...
                 // TODO: convert output file checks to a resource lock
-                if (!tryLockProjectFor(node)
-                    || !tryLockSharedResourceFor(node)
-                    || !workerLease.tryLock()
-                    || !canRunWithCurrentlyExecutedNodes(node, mutations)) {
+                } else if (!canRunWithCurrentlyExecutedNodes(node, mutations)) {
                     resourceLockState.releaseLocks();
+                    LOGGER.debug("Node {} cannot run with currently running nodes {}.", node, runningNodes);
                     continue;
                 }
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/execution/plan/DefaultExecutionPlan.java]: Can we avoid having to repeat `resourceLockState.releaseLocks()`? It seems too easy to forget to do it should a new case be added.

File: subprojects/core/src/main/java/org/gradle/execution/plan/DefaultExecutionPlan.java [hunk 2]
```diff
@@ -610,6 +614,7 @@ public Node selectNext(WorkerLeaseRegistry.WorkerLease workerLease, ResourceLock
         for (Iterator<Node> iterator = dependenciesWhichRequireMonitoring.iterator(); iterator.hasNext();) {
             Node node = iterator.next();
             if (node.isComplete()) {
+                LOGGER.debug("Monitored node {} completed. Updating dependencies.", node);
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/execution/plan/DefaultExecutionPlan.java]: Nit: using whole sentences in debug logs seems weird to me. I don't think we do it anywhere else.

File: subprojects/core/src/main/java/org/gradle/execution/plan/DefaultExecutionPlan.java [hunk 3]
```diff
@@ -645,6 +660,7 @@ public Node selectNext(WorkerLeaseRegistry.WorkerLease workerLease, ResourceLock
                 return node;
             }
         }
+        LOGGER.debug("No node could be selected, nodes ready: {}", foundReadyNode);
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/execution/plan/DefaultExecutionPlan.java]: Each of these logs add only a few nanoseconds, but the scheduler is a pretty hot place, too. I wonder how many times do we execute these log statements in a large build.

File: subprojects/core/src/main/java/org/gradle/execution/plan/DefaultExecutionPlan.java [hunk 4]
```diff
@@ -535,26 +510,20 @@ public void getNodeValues(Node node, Collection<? super Void> values, Collection
         if (cycles.isEmpty()) {
             // TODO: This isn't correct. This means that we've detected a cycle while determining the execution plan, but the graph walker did not find one.
             // https://github.com/gradle/gradle/issues/2293
-            throw new GradleException("Misdetected cycle between " + node + " and " + successor + ". Help us by reporting this to https://github.com/gradle/gradle/issues/2293");
+            throw new GradleException("Misdetected cycle between " + currentNode + " and " + successor + ". Help us by reporting this to https://github.com/gradle/gradle/issues/2293");
         }
         final List<Node> firstCycle = new ArrayList<>(cycles.get(0));
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/execution/plan/DefaultExecutionPlan.java]: Nit: unnecessary `final`.

File: subprojects/core/src/main/java/org/gradle/execution/plan/DefaultExecutionPlan.java (hunk 2 above)
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/execution/plan/DefaultExecutionPlan.java]: What would you expect as a log message here?

File: subprojects/core/src/main/java/org/gradle/execution/plan/DefaultExecutionPlan.java (hunk 3 above)
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/execution/plan/DefaultExecutionPlan.java]: We execute the log statements:
- each time a node completes
- the first time no ready node is found
- every time a lock changes and there are ready nodes

The lock changing does not happen to often, and most times no nodes are ready for execution. So logging (or checking if debug is enabled) should not happen too often.

File: subprojects/core/src/main/java/org/gradle/execution/plan/DefaultExecutionPlan.java (hunk 2 above)
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/execution/plan/DefaultExecutionPlan.java]: I changed it to not be two whole sentences.

File: subprojects/core/src/main/java/org/gradle/execution/plan/DefaultExecutionPlan.java (hunk 4 above)
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/execution/plan/DefaultExecutionPlan.java]: Removed.

File: subprojects/core/src/main/java/org/gradle/execution/plan/DefaultExecutionPlan.java (hunk 1 above)
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/execution/plan/DefaultExecutionPlan.java]: Done.

**Comment by wolfs** [comment]
@adammurdoch Can you check if the changes work with instant execution?

**Pull request merged**